File: deprecated/lib/output.ts
/**
 * Terminal output: colors, logging, and display helpers.
 */

// Color palette
export const COLORS: Record<string, string> = {
  cyan: "\x1b[36m",
  yellow: "\x1b[33m",
  magenta: "\x1b[35m",
  green: "\x1b[32m",
  blue: "\x1b[34m",
  orange: "\x1b[38;5;208m",
  red: "\x1b[31m",
  lavender: "\x1b[38;5;183m",
  pink: "\x1b[38;5;213m",
  teal: "\x1b[38;5;51m",
  lime: "\x1b[38;5;154m",
  coral: "\x1b[38;5;209m",
  sky: "\x1b[38;5;117m",
  gold: "\x1b[38;5;220m",
  violet: "\x1b[38;5;135m",
  reset: "\x1b[0m",
  dim: "\x1b[2m",
  bold: "\x1b[1m",
};

// Deterministic color palette for services (assigned by index)
// Order matters: avoid red/yellow early (error/warning associations)
// First ~10 should be distinct, non-alarming colors
export const SERVICE_COLORS = [
  "cyan",
  "yellow",
  "magenta",
  "teal",
  "green",
  "blue",
  "orange",
  "pink",
  "lavender",
  "violet",
  "lime",
  "coral",
  "sky",
  "gold",
  "violet",
];

// Disable colors when not a TTY (piping to other commands)
export const IS_TTY = Deno.stdout.isTerminal();

// ASCII key codes for cmdTop
export const KEY_Q_LOWER = 113;
export const KEY_Q_UPPER = 81;
export const KEY_CTRL_C = 3;

// Global verbose flag
let VERBOSE = false;

export function setVerbose(v: boolean): void {
  VERBOSE = v;
}

// Get color code only if TTY
export function c(color: string): string {
  if (!IS_TTY) return "";
  return COLORS[color] ?? "";
}

// Safe print that handles broken pipe (EPIPE) gracefully
export function print(msg: string): void {
  try {
    console.log(msg);
  } catch (e) {
    if (e instanceof Deno.errors.BrokenPipe) {
      Deno.exit(0);
    }
    throw e;
  }
}

export function log(msg: string, prefix?: string, color?: string): void {
  const now = new Date();
  const ts =
    now.toLocaleTimeString("en-US", {
      hour12: false,
      hour: "2-digit",
      minute: "2-digit",
      second: "2-digit",
    }) +
    "." +
    now.getMilliseconds().toString().padStart(3, "0");

  const colorCode = color ? c(color) : "";
  const prefixStr = prefix
    ? `${colorCode}${prefix.padEnd(12)}${c("reset")} `
    : "";

  print(`${c("dim")}${ts}${c("reset")} ${prefixStr}${msg}`);
}

export function logSystem(msg: string): void {
  log(msg, "rig", "bold");
}

export function logError(msg: string): void {
  // Log errors to stderr
  const now = new Date();
  const ts =
    now.getHours().toString().padStart(2, "0") +
    ":" +
    now.getMinutes().toString().padStart(2, "0") +
    ":" +
    now.getSeconds().toString().padStart(2, "0") +
    "." +
    now.getMilliseconds().toString().padStart(3, "0");
  console.error(`${c("dim")}${ts}${c("reset")} ${c("red")}${"rig".padEnd(12)}${c("reset")} ${msg}`);
}

export function logVerbose(msg: string): void {
  if (VERBOSE) {
    log(msg, "rig", "dim");
  }
}

/**
 * Strip ANSI control codes that would mess up log prefixes.
 * Preserves color codes but removes cursor movement, line clearing, etc.
 */
export function stripControlCodes(line: string): string {
  return line
    .replace(/\r/g, "")                     // Carriage return
    .replace(/\x1b\[\d*[ABCD]/g, "")        // Cursor movement (up/down/forward/back)
    .replace(/\x1b\[\d*;\d*[Hf]/g, "")      // Cursor position
    .replace(/\x1b\[\d*G/g, "")             // Cursor to column
    .replace(/\x1b\[\d*[JK]/g, "")          // Clear screen/line
    .replace(/\x1b\[\?25[lh]/g, "");        // Hide/show cursor
}

File: src/output.rs
use std::collections::HashMap;
use std::io::{self, IsTerminal, Write};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::LazyLock;

use chrono::Local;
use regex::Regex;

/// Global verbose flag
static VERBOSE: AtomicBool = AtomicBool::new(false);

/// ANSI color codes
static COLORS: LazyLock<HashMap<&'static str, &'static str>> = LazyLock::new(|| {
    let mut m = HashMap::new();
    m.insert("cyan", "\x1b[36m");
    m.insert("yellow", "\x1b[33m");
    m.insert("magenta", "\x1b[35m");
    m.insert("green", "\x1b[32m");
    m.insert("blue", "\x1b[34m");
    m.insert("orange", "\x1b[38;5;208m");
    m.insert("red", "\x1b[31m");
    m.insert("lavender", "\x1b[38;5;183m");
    m.insert("pink", "\x1b[38;5;213m");
    m.insert("teal", "\x1b[38;5;51m");
    m.insert("lime", "\x1b[38;5;154m");
    m.insert("coral", "\x1b[38;5;209m");
    m.insert("sky", "\x1b[38;5;117m");
    m.insert("gold", "\x1b[38;5;220m");
    m.insert("violet", "\x1b[38;5;135m");
    m.insert("reset", "\x1b[0m");
    m.insert("dim", "\x1b[2m");
    m.insert("bold", "\x1b[1m");
    m
});

/// Deterministic color palette for services (assigned by index)
pub const SERVICE_COLORS: &[&str] = &[
    "cyan", "yellow", "magenta", "teal", "green", "blue", "orange", "pink", "lavender", "violet",
    "lime", "coral", "sky", "gold",
];

/// Check if stdout is a TTY
pub fn is_tty() -> bool {
    io::stdout().is_terminal()
}

/// Get ANSI color code, returns empty string if not a TTY
pub fn c(color: &str) -> &str {
    if !is_tty() {
        return "";
    }
    COLORS.get(color).copied().unwrap_or("")
}

/// Get raw ANSI code (always, even when not TTY — for stderr)
pub fn c_raw(color: &str) -> &str {
    COLORS.get(color).copied().unwrap_or("")
}

pub fn set_verbose(v: bool) {
    VERBOSE.store(v, Ordering::Relaxed);
}

pub fn is_verbose() -> bool {
    VERBOSE.load(Ordering::Relaxed)
}

/// Safe print handling broken pipe gracefully
pub fn print(msg: &str) {
    if writeln!(io::stdout(), "{}", msg).is_err() {
        // Broken pipe — exit cleanly
        std::process::exit(0);
    }
}

/// Format a timestamp for log output
fn timestamp() -> String {
    Local::now().format("%H:%M:%S%.3f").to_string()
}

/// Log a message with optional prefix and color
pub fn log(msg: &str, prefix: &str, color: &str) {
    let ts = timestamp();
    let prefix_str = if !prefix.is_empty() {
        format!("{}{:<12}{} ", c(color), prefix, c("reset"))
    } else {
        String::new()
    };
    print(&format!(
        "{}{}{} {}{}",
        c("dim"),
        ts,
        c("reset"),
        prefix_str,
        msg
    ));
}

pub fn log_system(msg: &str) {
    log(msg, "rig", "bold");
}

pub fn log_error(msg: &str) {
    let ts = timestamp();
    let is_tty = io::stderr().is_terminal();
    let dim = if is_tty { c_raw("dim") } else { "" };
    let reset = if is_tty { c_raw("reset") } else { "" };
    let red = if is_tty { c_raw("red") } else { "" };
    let _ = writeln!(
        io::stderr(),
        "{}{}{} {}{:<12}{} {}",
        dim,
        ts,
        reset,
        red,
        "rig",
        reset,
        msg
    );
}

pub fn log_verbose(msg: &str) {
    if is_verbose() {
        log(msg, "rig", "dim");
    }
}

/// Strip ANSI control codes that would mess up log prefixes.
/// Preserves color codes but removes cursor movement, line clearing, etc.
pub fn strip_control_codes(line: &str) -> String {
    static RE_CR: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\r").unwrap());
    static RE_CURSOR_MOVE: LazyLock<Regex> =
        LazyLock::new(|| Regex::new(r"\x1b\[\d*[ABCD]").unwrap());
    static RE_CURSOR_POS: LazyLock<Regex> =
        LazyLock::new(|| Regex::new(r"\x1b\[\d*;\d*[Hf]").unwrap());
    static RE_CURSOR_COL: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\x1b\[\d*G").unwrap());
    static RE_CLEAR: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\x1b\[\d*[JK]").unwrap());
    static RE_CURSOR_VIS: LazyLock<Regex> =
        LazyLock::new(|| Regex::new(r"\x1b\[\?25[lh]").unwrap());

    let s = RE_CR.replace_all(line, "");
    let s = RE_CURSOR_MOVE.replace_all(&s, "");
    let s = RE_CURSOR_POS.replace_all(&s, "");
    let s = RE_CURSOR_COL.replace_all(&s, "");
    let s = RE_CLEAR.replace_all(&s, "");
    let s = RE_CURSOR_VIS.replace_all(&s, "");
    s.into_owned()
}

/// Strip all ANSI escape codes (for test assertions)
pub fn strip_ansi(s: &str) -> String {
    static RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\x1b\[[0-9;]*m").unwrap());
    RE.replace_all(s, "").into_owned()
}
